### ISSUE TITLE
some EMOJI shortcut not working

### ISSUE THREAD

**Issue by leizard** [Ranks/emojione#142]
Hi, thank you for great package
I'm using emojione in my Meteor project. it works great. But some emoji don't work. such as :spy: or :eye: .
Could you tell me why ?
Maybe the Meteor version is not the newest ?

**Comment by mikebe11** [comment]
I'm afraid I don't know the Meteor platform. Another user was kind enough to provide it.
I wonder if @qnub, @engelgabriel or @marceloschmidt would be able to assist here.

**Comment by qnub** [comment]
@leizard if you've used my `qnub:emojione` package — now you can update it to actual version `1.5.1` which contain latest emojis.

@mikebe11 thanks for notification!

**Comment by leizard** [comment]
thank you @qnub for your awesome work :+1:

**Comment by engelgabriel** [comment]
Hi @splendido do you know why our repo is considered "Not a Meteor pacakge"

![image](https://cloud.githubusercontent.com/assets/1000217/10700022/5318cd78-79bb-11e5-873a-81460a935916.png)

I am trying to get emojione into https://autopublish.meteor.com/

Can you help?

**Comment by engelgabriel** [comment]
@mikebe11 can you add me to the contributers list so I can finish configuring  https://autopublish.meteor.com/ ?

**Comment by mikebe11** [comment]
@engelgabriel Sure thing. I need some guidance here. Never set this up before. So "contributors list"? Is that the emojione organization member list in my Meteor account?

**Comment by engelgabriel** [comment]
@mikebe11 you'd have to go to: https://autopublish.meteor.com/accounts 
and do what is in the video: https://autopublish.meteor.com/howitworks

Or add me temporarily to your github project so I can do it for you.

![image](https://cloud.githubusercontent.com/assets/1000217/10710268/91ce9336-7a51-11e5-8567-22cf0ab63294.png)

**Comment by mikebe11** [comment]
@engelgabriel Was successful up to the point in the picture above. Clicked both sliders then around 40sec. into the video I couldn't go any further. I can't push anything. `git push origin 1.5.1` yields
```
error: src refspec 1.5.1 does not match any.
error: failed to push some refs to 'https://github.com/Ranks/emojione.git'
```

**Comment by splendido** [comment]
hey guys, nice work!
To enable autopublish for the package, toggling the `Autopublish` slider from the above picture should be enough.
Then, everytime a new release of the repo will be created, autopublish will try to publish the repo for Meteor.
Be aware that you need to add `publishbot` among the package maintainer with

```meteor admin maintainers emojione:emojione --add publishbot```


...I'm trying to have a look at this to see whether I can spot something weird and possibly solve the issue.
But perhaps @leizard could tell us which shortcut is not working for him?

**Comment by splendido** [comment]
@mikebe11 I'm sorry the video is a bit outdated: at that time the webhook was configured for tags, so every time a new tag was pushed to the github repo (```git tag 1.5.1 && git push origin 1.5.1```) the automatic publishing was attempted.

But now it works only for new releases

**Comment by splendido** [comment]
ok, #145 is my take on this.

Btw, I wouldn't disregard a different approach where we could do (note the triple braces...)

```handlebars
{{{emojione :smile:}}}
```

and this would be achieved with

```javascript
Template.registerHelper('emojione', function (str) {
    return emojione.toImage(str);
});
```

put inside `emojione-client.js`


This solution would still allow the use of other helper functions to get the shortcut string since now there's support for nested helpers:

```handlebars
{{{emojione (emoji)}}}
```

```javascript
Session.set('em', ':smile:');
Template.registerHelper("emoji", function(){
  return Session.get('em');
});
```

if you like it, feel free to have a look at #146

**Comment by splendido** [comment]
I can confirm that `:spy:` and `:eye:` seems to be working

**Comment by mikebe11** [comment]
Thanks @splendido for all the input. I can't add maintainers yet. It appears emojione:emojione is still not a valid package. Trying to list maintainers that's what I'm told plus it isn't listed on atmospherejs yet. Is there anything else I need to do for that initial publish?
As for the PRs we appreciate them very much. #145 is great. Will #146 be a suitable approach for the majority of Meteor users? Please excuse my ignorance here. We love being able to offer the package but until we have someone on staff who knows the platform we rely on you Meteor users/experts to advise us. If you can't use `{{#emojione}}My emoji :beers: text.{{/emojione}}` would users still be able to parse blocks of text with #146. Would it now be something like 
`{{{emojione (My emoji :beers: text.)}}}`
or
`My emoji {{{emojione :beers:}} text.` ?

**Comment by splendido** [comment]
@mikebe11 I see. Just for me to catchup with the current status of works and understand what is still to be done, here are the steps I'd suggest to publish a package:

### 1. user and organization set-up:

1. get a user on [Meteor.com](https://www.meteor.com/) ('sign-in' on the top right and then 'register'): this could be `mikebe11`
2. create an new organization from the _user -> organizations_ page to reserve a namespace for your packages: `emojione` should be just fine here
3. (optional) possibly add other collaborators to the organization (to let them publish packages of mange the organization for you) clicking the _edit members_ link on the top right corner of the organization block

### 2. (first time) package publishing:

1. install meteor locally with `curl https://install.meteor.com/ | sh` (OS X or Linux)
2. get a local clone of the package repo
3. publish the new package (first time only) running `meteor publish --create` from inside the rpository's root folder, for successive release `meteor publish` should be used
4. (optional) add other package maintainers to the newly created package with: `meteor admin maintainers emojione:emojione --add <username>` so that others can publish upcoming releases

packages' name are always in the form _namespace:pkgname_ where _namespace_ can be both a username or an organization name.
Please note that for a package which _namespace_ is a username there's no way to add maintainers before it is published for the first time.

### 3. autopublish.meteor.com set-up

9. add `publishbot` to the package maintainers with `meteor admin maintainers emojione:emojione --add publishbot`
10. go to autopublish.meteor.com and toggle the `Autopublish` slider for the repository containing the package (already done!)

every new github release will be automatically published by the bot user.

Note here that in case `emojione` was created as an organization and `publishbot` was added among its members a new release would correctly trigger the first publish for the package without need to install meteor locally to create the package _by hands_

**Comment by splendido** [comment]
with #146 the following would work:

```handlebars
{{{emojione 'My emoji :beers: text.'}}}
My emoji {{{emojione :beers:}}} text.
```

In case of some message text coming from the DB, something like this could be used:

```handlebars
{{{emojione (msg) }}}
```

where `msg` is another helper function returning the whole text for the message to be rendered (```My emoji :beers: text.``` for the above case)

I think it's more a thing about personal preferences. Moreover triple braces might be harmful if the text passed in is not correctly sanitazed (which should be explicitely done by the developer wihin the `msg` helper in the example above.). Sanitization should probably be performed anyway before directly exposing messages coming from the DB...

**Comment by mikebe11** [comment]
@splendido Excellent. Most was already done. I did the publish --create and that looks like all that was needed. I also added publishbot in my account organization and via command line. Package is now on atmospherejs. Thanks so much!

@qnub , @engelgabriel Any thoughts/preference on splendido's contribution? #145 or #146?

**Comment by splendido** [comment]
:+1:

**Comment by mikebe11** [comment]
@leizard  Someone's confirmed those shortnames are working. Are things up to date and working on your end?

**Comment by splendido** [comment]
@mikebe11 I've just added a commit to #145 to get the README.me correctly shown on atmosphere.
While waiting to decide what to do with the PR, you could cherry-pick e345c61, f82b7df, and e345c61 into master to get the updated documentation for Meteor (in particular, package name...).
You should then run `meteor publish --update` to get the new version of the README onto Atmosphere

**Comment by mikebe11** [comment]
Ok, those 3 readme commits have been merged. Plus the package update.

**Comment by splendido** [comment]
@qnub: what about deprecating your package now that `emojione:emojione` is up onto the atmosphere and running?

`meteor admin set-unmigrated qnub:emojione`

**Comment by qnub** [comment]
@splendido done. also i've added notification to description and README.

**Comment by qnub** [comment]
But not sure that aoutopublication to atmosphere is working. Here i can see 1.5.2 version, but in atmosphere still 1.5.1…

**Comment by mikebe11** [comment]
Can see here that it tried to autopublish but there was an error.
https://autopublish.meteor.com/

**Comment by splendido** [comment]
I still hadn't figured out why sometimes there's very little free memory on the build machine
Let me try to fix it...

sorry

**Comment by splendido** [comment]
ok, now worked but the new version was already published (I guess by hands by @mikebe11 )
The problem was the update to Meteor 1.2.1 from last 10/27: automatic updates always fails :-(
Now it should be good to go!

**Comment by mikebe11** [comment]
Yeah, i did try to publish through the command line when I saw the autopub didn't go through.
Thanks

**Comment by engelgabriel** [comment]
I am back from holiday. Let me take a better look at this tomorrow.

**Comment by engelgabriel** [comment]
All looks good!!! :) We are using the official package now.

What are your thoughts on how to make the package dependence on cdn optional?
https://github.com/RocketChat/Rocket.Chat/issues/1046 

Maybe have a flag to use the SVG Sprites?
https://github.com/Ranks/emojione/blob/master/assets/sprites/emojione.sprites.svg

**Comment by mikebe11** [comment]
That's great!

We already have the following options available to change the image paths and to use SVGs or SVG sprites. Does Meteor allow for these to be easily set after install?

```javascript
    ns.imagePathPNG = '//cdn.jsdelivr.net/emojione/assets/png/';
    ns.imagePathSVG = '//cdn.jsdelivr.net/emojione/assets/svg/';
    ns.imagePathSVGSprites = './../assets/sprites/emojione.sprites.svg';
    ns.imageType = 'png'; // or svg
    ns.sprites = false; // if this is true then sprite markup will be used (if SVG image type is set then you must include the SVG sprite file locally)
```

**Comment by engelgabriel** [comment]
Given the file size of the SVG sprite, I'd only add the PNG file to the meteor "assets" folder by default.

We can create environment variable like:
```
EMOJIONE_ADD_SVG_SPRITES = false
EMOJIONE_ADD_PNG_SPRITES = true
```
To control the meteor build behaviour on the package.js

**Comment by mikebe11** [comment]
@engelgabriel Adding environment variables, like you've mentioned, to package.js seems reasonable.
Can the Meteor build also have options for other users to stick to the CDN? It's great to be flexible for RocketChat but if other package users wish to stick to the CDN we'd like them to be able to do so.

Also, a heads up, the sheets for the new collection will be smaller, PNG ~2MB and SVG ~4.5MB

**Comment by engelgabriel** [comment]
Yes, we can put all config as possible environment variables that can overwrite the default values. That would be the best way for maximum flexibility.

**Comment by mikebe11** [comment]
Sounds perfect, flexible and maintains separation from the actual js file. If you can submit a PR on that that'd be great or contact someone to submit it.

**Comment by engelgabriel** [comment]
I'll do it this week.

**Comment by leizard** [comment]
@mikebe11  I can confirm all this work great and updated now in Meteor :)

**Comment by engelgabriel** [comment]
@marceloschmidt just submitted the PR #159 that will add the options to use sprites :+1:

**Comment by mikebe11** [comment]
@marceloschmidt  Thank you very much for the update to package.js. It's been merged.
@engelgabriel  Thank you too for the Meteor guidance. Our updated library and new emoji collection will be pushed very soon. We hope it's well received and look forward to continued interaction with the Meteor community.
This original issue is resolved and we got #159 merged so I'm gonna close this issue. 
Cheers
